modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPClusterSubscriptionHandler.java

```diff
@@ -0,0 +1,211 @@
+
+        // Check topic subscriptions
+        subscriptionAvailable = topicSubscriptionHandler.isSubscriptionAvailable(subscription);
+
+        // Check durable topic subscriptions
+        if (!subscriptionAvailable) {
+            subscriptionAvailable = durableTopicSubscriptionHandler.isSubscriptionAvailable(subscription);
+        }
+
+        // Check queue subscriptions
+        if (!subscriptionAvailable) {
+            for (Map.Entry<String, Set<AndesSubscription>> entry : queueSubscriptionMap.entrySet()) {
+                if (entry.getValue().contains(subscription)) {
+                    subscriptionAvailable = true;
+                    break;
+                }
+            }
+        }
+
+        return subscriptionAvailable;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void removeSubscription(AndesSubscription subscription) {
+        if (DestinationType.TOPIC == subscription.getDestinationType()) {
+            topicSubscriptionHandler.removeSubscription(subscription);
+        } else if (DestinationType.DURABLE_TOPIC == subscription.getDestinationType()) {
+            durableTopicSubscriptionHandler.removeSubscription(subscription);
+        } else {
+            removeQueueSubscriptions(subscription);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<AndesSubscription> getMatchingSubscriptions(String destination, DestinationType destinationType) {
+        Set<AndesSubscription> matchingSubscriptions;
+        if (DestinationType.TOPIC == destinationType) {
+            matchingSubscriptions = topicSubscriptionHandler.getMatchingSubscriptions(destination, destinationType);
+        } else if (DestinationType.DURABLE_TOPIC == destinationType) {
+            matchingSubscriptions = durableTopicSubscriptionHandler.getMatchingSubscriptions(destination,
+                    destinationType);
+        } else {
+            matchingSubscriptions = queueSubscriptionMap.get(destination);
+        }
+
+        if (null == matchingSubscriptions) {
+            matchingSubscriptions = Collections.emptySet();
+        }
+        return matchingSubscriptions;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<AndesSubscription> getAllSubscriptions() {
+        List<AndesSubscription> subscriptionList = new ArrayList<>();
+
+        subscriptionList.addAll(topicSubscriptionHandler.getAllSubscriptions());
+
+        subscriptionList.addAll(durableTopicSubscriptionHandler.getAllSubscriptions());
+
+        for (Map.Entry<String, Set<AndesSubscription>> entry : queueSubscriptionMap.entrySet()) {
+            subscriptionList.addAll(entry.getValue());
+        }
+
+        return subscriptionList;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<String> getAllDestinations(DestinationType destinationType) {
+        Set<String> destinations;
+
+        if (DestinationType.TOPIC == destinationType) {
```

Comment:
Rather than checking whether it is topic or durable topic always in each method, cant we decide the subscription path in an upper level? So that topic subscription goes in one path and durable subscriptions goes in another path?